Add support for shared channels

ericdwang opened the pull request:
Since shared channels only work with the new conversations API, a new
class of SlackChannel and API type were added to support them. Also
`rtm.start` doesn't include external users so their information needs to
be fetched individually when getting the channel history.

To distinguish external users from team members their nicks get appended
with an asterisk, which can be changed with a new setting. They also
appear in a new "External" group in the nicklist since their presences
can't be fetched with `users.getPresence`.

Unfortunately `conversations.history` doesn't include thread messages so
threads won't show up in shared channels.

Sidenote: the asterisk suffix is also completely arbitrary so I'm open to any suggestions if it doesn't make sense!

auscompgeek commented:
Slack uses a different icon to signify shared channels in their apps, correct? Would it be worth allowing the user to change the prefix of shared channels as well, similarly to private channels?

ericdwang commented:
They actually use "#" for shared channels too, but they're listed separately under a "Shared Channels" section. I think it'd still be good to distinguish shared channels though, so I changed the prefix to "%" and added a setting for it. Thanks for the idea! Open to suggestions for this default too.

auscompgeek commented:
Hmm. It's tempting to keep the default as `#` then. I'll let others weigh in. Picking sensible defaults is hard.

Out of curiosity, how does this handle channels (both public and private) becoming shared channels? (Speaking of, I wonder how we handle public channels becoming private. Hmm…)

ericdwang commented:
There's also an icon to the right of shared channels which normal ones don't have:

![slack](https://user-images.githubusercontent.com/2069450/32141487-0bf27778-bc3f-11e7-8ec6-55ab749cb89d.png)

And uhh I'm not sure what happens there, unfortunately I can't test it because I'm not an admin of my team...

ericdwang commented:
This was broken for a while after Slack changed `rtm.start` so that members weren't included in shared channels anymore for some reason, so I pushed a fix for that. I just added the requests where things seemed to work, not sure what the actual best places would be for them.

denibertovic commented:
I tried testing this but got the following traceback:

```
2018-01-25 15:10:04		python: stdout/stderr: Traceback (most recent call last):
2018-01-25 15:10:04		python: stdout/stderr:   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 510, in handle_next
2018-01-25 15:10:04		python: stdout/stderr:     EVENTROUTER.handle_next()
2018-01-25 15:10:04		python: stdout/stderr:   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 498, in handle_next
2018-01-25 15:10:04		python: stdout/stderr:     self.handlers[function_name](j, self, **kwargs)
2018-01-25 15:10:04		python: stdout/stderr:   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 2171, in handle_rtmstart
2018-01-25 15:10:04		python: stdout/stderr:     channels[item["id"]] = SlackDMChannel(eventrouter, users, **item)
2018-01-25 15:10:04		python: stdout/stderr:   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 1626, in __init__
2018-01-25 15:10:04		python: stdout/stderr:     kwargs["name"] = users[dmuser].name
2018-01-25 15:10:04		python: stdout/stderr: KeyError: '****redacted_user_id****'
```

Initially I thought that it has to do with there being open DM channels with folks from the shared channel, so I closed those from the web interface in the hopes that they won't pop up over the api as well... but it didn't help.

denibertovic commented:
@trygveaa @ericdwang  This is failing specifically because of DM's (SlackDMChannel) with users from a shared channel. I'm unable to figure out a way to make it work .... other than completely ignoring those DMs but that's not really helpful.

tonyskapunk commented:
Tested this one, and got same results as @denibertovic, in my case I have a DM with a user in a shared channel.

```
09:12:31  python: stdout/stderr: Traceback (most recent call last):
09:12:31  python: stdout/stderr:   File "/home/tonyskapunk/.weechat/python/autoload/wee_slack.py", line 510, in handle_ne
09:12:31  python: stdout/stderr:     EVENTROUTER.handle_next()
09:12:31  python: stdout/stderr:   File "/home/tonyskapunk/.weechat/python/autoload/wee_slack.py", line 498, in handle_ne
09:12:31  python: stdout/stderr:     self.handlers[function_name](j, self, **kwargs)
09:12:31  python: stdout/stderr:   File "/home/tonyskapunk/.weechat/python/autoload/wee_slack.py", line 2171, in handle_rt
09:12:31  python: stdout/stderr:     channels[item["id"]] = SlackDMChannel(eventrouter, users, **item)
09:12:31  python: stdout/stderr:   File "/home/tonyskapunk/.weechat/python/autoload/wee_slack.py", line 1626, in __init__ 
09:12:31  python: stdout/stderr:     kwargs["name"] = users[dmuser].name
09:12:31  python: stdout/stderr: KeyError: u'**ID of user in shared channel**'
```

Linuus commented:
What is the status of this PR? Does it work? :)

We use shared channels at work and I’d really like to try out wee-slack! Thank you for the great work!

❤️

denibertovic commented:
@trygveaa is is possible to pay for the development of this feature? what would that look like?

ericdwang commented:
I apologize for not updating or responding to this in so long... I don't have any DMs with external users (I didn't even know it was possible) so it never bothered me and I didn't have a way to test it. Based on the tracebacks provided I pushed a fix that may work (tested with regular DMs, not external users) which you can try out. Not sure if getting history for them will work either, that may require switching to the `conversations.history` method.

Some other things have broken since the last update (shared file links, External nicklist group, etc.) so I've fixed those as well. Slack keeps changing their API... Also rebased, fixed conflicts, and squashed some old commits.

Linuus commented:
Awesome! I’ll need to try this out. Thanks!

denibertovic commented:
Hi @ericdwang , thank you for the update! I'm testing the changes out and I can confirm that I'm able to join the slack teams that have shared channels. I'm able to join those shared channels and see I'm able to see DM's with users from shared channels. 

The only thing I'm not seeing is I'm not seeing the list of all external users in a shared channel. I'm guessing the traceback below has something to do with it.

```
2018-05-15 16:25:57		python: stdout/stderr (weemoji): Traceback (most recent call last):
2018-05-15 16:25:57		python: stdout/stderr (slack): Traceback (most recent call last):
2018-05-15 16:25:57		python: stdout/stderr (weemoji):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 126, in wrapper
2018-05-15 16:25:57		python: stdout/stderr (slack):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 126, in wrapper
2018-05-15 16:25:57		python: stdout/stderr (slack):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 126, in wrapper
2018-05-15 16:25:57		python: stdout/stderr (weemoji):     return f(*decode_from_utf8(args), **decode_from_utf8(kwargs))
2018-05-15 16:25:57		python: stdout/stderr (slack):     return f(*decode_from_utf8(args), **decode_from_utf8(kwargs))
2018-05-15 16:25:57		python: stdout/stderr (weemoji):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 741, in buffer_list_update_callback
2018-05-15 16:25:57		python: stdout/stderr (slack):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 741, in buffer_list_update_callback
2018-05-15 16:25:57		python: stdout/stderr (slack):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 741, in buffer_list_update_callback
2018-05-15 16:25:57		python: stdout/stderr (weemoji):     b[1].refresh()
2018-05-15 16:25:57		python: stdout/stderr (slack):     b[1].refresh()
2018-05-15 16:25:57		python: stdout/stderr (weemoji):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 1741, in refresh
2018-05-15 16:25:57		python: stdout/stderr (slack):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 1741, in refresh
2018-05-15 16:25:57		python: stdout/stderr (slack):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 1741, in refresh
2018-05-15 16:25:57		python: stdout/stderr (weemoji):     return self.rename()
2018-05-15 16:25:57		python: stdout/stderr (slack):     return self.rename()
2018-05-15 16:25:57		python: stdout/stderr (weemoji):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 1733, in rename
2018-05-15 16:25:57		python: stdout/stderr (slack):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 1733, in rename
2018-05-15 16:25:57		python: stdout/stderr (slack):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 1733, in rename
2018-05-15 16:25:57		python: stdout/stderr (weemoji):     new_name = self.formatted_name(style="sidebar", present=self.team.is_user_present(self.user), enable_color=config.colorize_private_chats)
2018-05-15 16:25:57		python: stdout/stderr (slack):     new_name = self.formatted_name(style="sidebar", present=self.team.is_user_present(self.user), enable_color=config.colorize_private_chats)
2018-05-15 16:25:57		python: stdout/stderr (weemoji):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 1108, in is_user_present
2018-05-15 16:25:57		python: stdout/stderr (slack):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 1108, in is_user_present
2018-05-15 16:25:57		python: stdout/stderr (slack):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 1108, in is_user_present
2018-05-15 16:25:57		python: stdout/stderr (weemoji):     if user.presence == 'active':
2018-05-15 16:25:57		python: stdout/stderr (slack):     if user.presence == 'active':
2018-05-15 16:25:57		python: stdout/stderr (weemoji): AttributeError: 'NoneType' object has no attribute 'presence'
2018-05-15 16:25:57		python: stdout/stderr (slack): AttributeError: 'NoneType' object has no attribute 'presence'
2018-05-15 16:25:57	=!=	python: error in function "buffer_list_update_callback"
```

trygveaa commented on wee_slack.py:
```diff
@@ -2270,6 +2351,30 @@ def handle_history(message_json, eventrouter, **kwargs):
+    team = eventrouter.teams[request_metadata.team_hash]
+    channel = team.channels[request_metadata.channel_identifier]
+    user_info = user_json['user']
+    user_info.update(is_external=True, deleted=False)
```
I'm not sure we should assume all `users.info` calls are for external users. Even though they are now, we might use the method for other things in the future.

trygveaa commented on wee_slack.py:
```diff
@@ -1634,12 +1662,21 @@ class SlackDMChannel(SlackChannel):
+        super(SlackDMChannel, self).set_related_server(team)
+        # If the topic hasn't been set then the user is external and needs to
+        # be fetched
+        if not self.topic:
```
It's a bit fragile and weird to rely on the topic not being set. Instead you can rather check if `self.user not in self.team.users`.

trygveaa commented on wee_slack.py:
```diff
@@ -982,6 +990,7 @@ def __init__(self, eventrouter, token, websocket_url, subdomain, nick, myidentif
         except:
             self.channels = channels
         self.users = users
+        self.external_users = {}
```
Is there a reason for creating a separate list for external users instead of placing them in the normal users list?

trygveaa commented:
@ericdwang: I fixed the things I commented on, and pushed to this PR. Hope that's okay :)

I didn't manage to reproduce the `KeyError: u'channel'` error I got when I opened the shared DM. Now the response did contain the channel attribute, so I didn't change anything there. Otherwise, I have fixed the rest of the things I mentioned.

denibertovic commented:
```
2018-06-11 14:09:28             python: stdout/stderr (slack): Traceback (most recent call last):
2018-06-11 14:09:28             python: stdout/stderr (slack):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 517, in handle_next
2018-06-11 14:09:28             python: stdout/stderr (slack):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 517, in handle_next
2018-06-11 14:09:28             python: stdout/stderr (slack):     EVENTROUTER.handle_next()
2018-06-11 14:09:28             python: stdout/stderr (slack):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 507, in handle_next
2018-06-11 14:09:28             python: stdout/stderr (slack):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 507, in handle_next
2018-06-11 14:09:28             python: stdout/stderr (slack):     raise ProcessNotImplemented(function_name)
2018-06-11 14:09:28             python: stdout/stderr (slack): ProcessNotImplemented: desktop_notification
2018-06-11 14:09:28             python: stdout/stderr (slack): Traceback (most recent call last):
2018-06-11 14:09:28             python: stdout/stderr (slack):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 517, in handle_next
2018-06-11 14:09:28             python: stdout/stderr (slack):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 517, in handle_next
2018-06-11 14:09:28             python: stdout/stderr (slack):     EVENTROUTER.handle_next()
2018-06-11 14:09:28             python: stdout/stderr (slack):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 507, in handle_next
2018-06-11 14:09:28             python: stdout/stderr (slack):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 507, in handle_next
2018-06-11 14:09:28             python: stdout/stderr (slack):     raise ProcessNotImplemented(function_name)
2018-06-11 14:09:28             python: stdout/stderr (slack): ProcessNotImplemented: conversationsinfo
2018-06-11 14:09:28             python: stdout/stderr (slack): Traceback (most recent call last):
2018-06-11 14:09:28             python: stdout/stderr (slack):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 517, in handle_next
2018-06-11 14:09:28             python: stdout/stderr (slack):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 517, in handle_next
2018-06-11 14:09:28             python: stdout/stderr (slack):     EVENTROUTER.handle_next()
2018-06-11 14:09:28             python: stdout/stderr (slack):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 507, in handle_next
2018-06-11 14:09:28             python: stdout/stderr (slack):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 507, in handle_next
2018-06-11 14:09:28             python: stdout/stderr (slack):     raise ProcessNotImplemented(function_name)
2018-06-11 14:09:28             python: stdout/stderr (slack): ProcessNotImplemented: conversationsinfo
2018-06-11 14:09:31             python: stdout/stderr (slack): Traceback (most recent call last):
2018-06-11 14:09:31             python: stdout/stderr (slack):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 517, in handle_next
2018-06-11 14:09:31             python: stdout/stderr (slack):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 517, in handle_next
2018-06-11 14:09:31             python: stdout/stderr (slack):     EVENTROUTER.handle_next()
2018-06-11 14:09:31             python: stdout/stderr (slack):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 505, in handle_next
2018-06-11 14:09:31             python: stdout/stderr (slack):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 505, in handle_next
2018-06-11 14:09:31             python: stdout/stderr (slack):     self.handlers[function_name](j, self, **kwargs)
2018-06-11 14:09:31             python: stdout/stderr (slack):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 2284, in handle_channelsinfo
2018-06-11 14:09:31             python: stdout/stderr (slack):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 2284, in handle_channelsinfo
2018-06-11 14:09:31             python: stdout/stderr (slack):     channel.set_unread_count_display(channel_json['channel']['unread_count_display'])
2018-06-11 14:09:31             python: stdout/stderr (slack): KeyError: u'unread_count_display'
2018-06-11 14:09:31             python: stdout/stderr (slack): Traceback (most recent call last):
2018-06-11 14:09:31             python: stdout/stderr (slack):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 517, in handle_next
2018-06-11 14:09:31             python: stdout/stderr (slack):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 517, in handle_next
2018-06-11 14:09:31             python: stdout/stderr (slack):     EVENTROUTER.handle_next()
2018-06-11 14:09:31             python: stdout/stderr (slack):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 505, in handle_next
2018-06-11 14:09:31             python: stdout/stderr (slack):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 505, in handle_next
2018-06-11 14:09:31             python: stdout/stderr (slack):     self.handlers[function_name](j, self, **kwargs)
2018-06-11 14:09:31             python: stdout/stderr (slack):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 2284, in handle_channelsinfo
2018-06-11 14:09:31             python: stdout/stderr (slack):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 2284, in handle_channelsinfo
2018-06-11 14:09:31             python: stdout/stderr (slack):     channel.set_unread_count_display(channel_json['channel']['unread_count_display'])
2018-06-11 14:09:31             python: stdout/stderr (slack): KeyError: u'unread_count_display'
2018-06-11 14:10:42             python: stdout/stderr (slack): Traceback (most recent call last):
2018-06-11 14:10:42             python: stdout/stderr (slack):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 517, in handle_next
2018-06-11 14:10:42             python: stdout/stderr (slack):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 517, in handle_next
2018-06-11 14:10:42             python: stdout/stderr (slack):     EVENTROUTER.handle_next()
2018-06-11 14:10:42             python: stdout/stderr (slack):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 507, in handle_next
2018-06-11 14:10:42             python: stdout/stderr (slack):   File "/home/deni/.weechat/python/autoload/wee_slack.py", line 507, in handle_next
2018-06-11 14:10:42             python: stdout/stderr (slack):     raise ProcessNotImplemented(function_name)
2018-06-11 14:10:42             python: stdout/stderr (slack): ProcessNotImplemented: member_joined_channel
```

@trygveaa @ericdwang  This is the output I'm seeing with the latest changes. I should note though that I am now correctly seeing users in the shared channel user list as well as correct names in the DM channels with external users I already had opened. 

However, for some reasons DM channels with external users are not showing up in my "quick launch" fuzzy find bar. I'm using `go.py` script to easily jump across multiple channels/dms and I can do that with normal slack channels and slack users just not with external users. I can, of course, navigate to those DM's using Alt+<left/right arrow> but that takes longer. I think this used to work btw...

trygveaa commented:
@denibertovic: Hm, that error is strange. It says it can't find `unread_count_display` in the response from `channels.info`. However, the documentation says it should be there. It is present for me, and I haven't seen that error here.

Can you post the response you get when that error appears? E.g. with this patch:

```
diff --git a/wee_slack.py b/wee_slack.py
index 2682ea9..f6b3d62 100644
--- a/wee_slack.py
+++ b/wee_slack.py
@@ -2281,8 +2281,12 @@ def handle_channelsinfo(channel_json, eventrouter, **kwargs):
     request_metadata = pickle.loads(channel_json["wee_slack_request_metadata"])
     team = eventrouter.teams[request_metadata.team_hash]
     channel = team.channels[request_metadata.channel_identifier]
-    channel.set_unread_count_display(channel_json['channel']['unread_count_display'])
-    channel.set_members(channel_json['channel']['members'])
+    try:
+        channel.set_unread_count_display(channel_json['channel']['unread_count_display'])
+        channel.set_members(channel_json['channel']['members'])
+    except KeyError:
+        from pprint import pformat
+        w.prnt('', pformat(channel_json['channel']))
 
 
 def handle_groupsinfo(group_json, eventrouter, **kwargs):
```


> However, for some reasons DM channels with external users are not showing up in my "quick launch" fuzzy find bar.

I guess this is because the buffer name is the ID of the DM channel instead of the nick of the person it belongs to. I see that it is named with the nick in the buflist (the list on the side), but is using the ID in the output of `/buffer` and in go.py.

denibertovic commented:
Hey @trygveaa , I can't paste the output because it's too long as has a lot of user id's in it but I don't see "unread_count_display" in there at all. I've checked a couple of time. It's worth nothing that the output was not caused by the team that has shared channels, or by joining a shared channel. It's from one of the OSS teams I've joined (https://fpchat-invite.herokuapp.com/). 

I think that the shared channels are raising `ProcessNotImplemented: member_joined_channel` because it appeared when I joined the shared channel, but that could be a coincidence.

As for the nick list error. I too see the correct nicks i the buffer list on the left side. But it would be nice if we could somehow name the buffers with the correct names so that switching between them works normally. Is that possible?

trygveaa commented:
> It's worth nothing that the output was not caused by the team that has shared channels, or by joining a shared channel. It's from one of the OSS teams I've joined (https://fpchat-invite.herokuapp.com/).

Aha, that's very useful to know. I checked that team, and there I get the error too. I'll look into it, but it's unrelated to this PR.

> I think that the shared channels are raising ProcessNotImplemented: member_joined_channel because it appeared when I joined the shared channel, but that could be a coincidence.

You can ignore the ProcessNotImplemented errors, they're not actually errors. With #589 which I just merged, these are not shown anymore, just logged to the debug buffer.

> As for the nick list error. I too see the correct nicks i the buffer list on the left side. But it would be nice if we could somehow name the buffers with the correct names so that switching between them works normally. Is that possible?

It's possible, I just haven't looked into it yet.

denibertovic commented:
> Aha, that's very useful to know. I checked that team, and there I get the error too. I'll look into it, but it's unrelated to this PR.

Yep, agreed.

> You can ignore the ProcessNotImplemented errors, they're not actually errors. With #589 which I just merged, these are not shown anymore, just logged to the debug buffer.


Awesome, thank you!

> It's possible, I just haven't looked into it yet.

Awesome! I'd say this is the only thing left then. But if you feel like merging the PR without it I'm okay with it. We can always fix it later. Let me know if there's something I can help with here.

ericdwang commented:
> @ericdwang: I fixed the things I commented on, and pushed to this PR. Hope that's okay :)

Sorry for the super late reply... no problem at all with me, thanks for fixing the issues and please fix any others you find! I unfortunately don't have access to a shared channel anymore so I won't be much help now anyway.
